Bug 1109675 - Implement 'Request Desktop Site'

Johennes opened the pull request.

Johennes commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -663,8 +663,15 @@ class BrowserViewController: UIViewController {
+        }
+
+        if #available(iOS 9.0, *) {
+            tab.webView?.customUserAgent = nil
```
This will reset the spoofed user agent whenever a new URL is entered manually or a link is selected from the home panel.

thebnich replied:
I think we want this to also include navigation from the page (clicking links, submitting forms, etc.), so I don't think `finishEditingAndSubmit` will be a wide enough net for what we need. How about moving this to `didCommitNavigation`?

Johennes commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -1020,6 +1027,13 @@ extension BrowserViewController: BrowserToolbarDelegate {
 
+    func browserToolbarDidLongPressReload(browserToolbar: BrowserToolbarProtocol, button: UIButton) {
+        if #available(iOS 9.0, *) {
+            tabManager.selectedTab?.webView?.customUserAgent = "Macintosh; Max OS 10_10"
```
This should probably be changed to something else but I wasn't sure to what.

thebnich replied:
So our iOS UA is:
Mozilla/5.0 (iPhone; CPU iPhone OS 8_3 like Mac OS X) AppleWebKit/**WebKitVersion** (KHTML, like Gecko) FxiOS/1.0 Mobile/12F69 Safari/**WebKitVersion**

For desktop, let's try this:
Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_1) AppleWebKit/**WebKitVersion** (KHTML, like Gecko) Safari/**WebKitVersion**

In other words, replacing the iPhone string with the Macintosh one and removing the FxiOS and Mobile parts of the UA. I'd put this in a static `desktopUserAgent` function in `UserAgent.swift`, right alongside our existing `defaultUserAgent`.

Johennes commented:
One thing I noticed while testing is that some sites redirect mobile user agents to dedicated subdomains like for instance m.facebook.com. Reloading that particular URL with a desktop user agent brings up a different but still mobile-targeted page instead of the expected desktop site. We could strip the m. prefix but I wasn't sure if that wouldn't be too much magic.

thebnich commented:
Hm...I wonder if `webView.backForwardList.currentItem?.initialURL` will give us the pre-redirect URL in these cases? If so, it might be worth considering using `loadRequest` with that URL instead of `reload`.

Johennes commented on Shared/UserAgent.swift:
```diff
@@ -85,4 +85,26 @@ public class UserAgent {
+        userAgent.replaceCharactersInRange(platformMatch.range, withString: "(Macintosh; Intel Mac OS X 10_11_1)")
+
+        // Strip mobile section
+        let mobileRegex = try! NSRegularExpression(pattern: " FxiOS/[^ ]+ Mobile/[^ ]+", options: [])
```
It feels a bit weird to strip the FxiOS section here because it is added in `defaultUserAgent` which is called before. Would it be worth to extract a private method with the common parts of `defaultUserAgent` and `desktopUserAgent`? If so, I think we'd also have to adapt the caching.

thebnich replied:
It does feel a little weird to add it only to strip it out again, but IMO not worth any refactoring changes for `defaultUserAgent` (I agree that we'd need to change what we cache).

Johennes commented:
Using the `backForwardList` for acquiring the initial URL seems to work. Great tip! I didn't know that we have the redirect history at our fingertips.

I also incorporated the other changes you suggested (user agent calculation and user agent reset on navigation). During testing in the simulator I - once in a while - see cases in which the long-press reload does not load the desktop site. It could be that the user agent resetting in `didCommitNavigation` undoes the spoofing too early in some case. I'll have to test this more thoroughly to be sure, though.

thebnich commented:
> It could be that the user agent resetting in didCommitNavigation undoes the spoofing too early in some case.

`didCommitNavigation` is called when the content *starts* to arrive, so that might not actually be the best callback to use. Sorry for suggesting the wrong method! `didFinishNavigation` sounds more appropriate -- maybe give that one a try?

Johennes commented:
I found some time for further testing and it seems that my initial suspicion was a bit hasty. Resetting the agent in `didCommitNavigation` **does** appear to be fine. The real problem was that `WKWebView`'s `loadRequest` method has some sort of internal caching logic and does _not always_ perform actual network requests when the URL to load is identical to the current URL. I verified this by monitoring my network traffic while running the app in the simulator. To overcome this, I switched to calling `reload` when the URL is not changed due to redirects and it appears to work as expected now.

One more thing I noticed about redirects and `m.` subdomains: Using the initial URL works great when you entered the desktop URL (like `facebook.com`) and were then redirected to a mobile subdomain (like `m.facebook.com`). However, once you **are** on the mobile subdomain and start navigating to different pages on that domain there is no redirection anymore. Hence, when you long-tap the reload button to request the desktop site, we cannot determine any initial URL. I'm not sure if we can (and should) work around this, however.

thebnich commented:
> To overcome this, I switched to calling reload when the URL is not changed due to redirects and it appears to work as expected now.

Note that `NSURLRequest` also has a `cachePolicy` initializer argument where you could pass `NSURLRequestReloadIgnoringLocalAndRemoteCacheData`. I think this is preferable to a separate reload so that we handle all loads consistently.

Johennes commented:
Yes, sorry, I forgot to mention that I also tried the `cachePolicy` parameter but to no avail. I just double checked this and for whatever reason it appears to only happen on `stackoverflow.com`. When I load that page and long-tap reload, I usually end up with the mobile site instead of the expected desktop site and there are no observable network requests during the reload. Only when I repeatedly long-tap reload, I ultimately see a network request and the page reloads with the desktop appearance. This seems to happen regardless of what I specify for the cache policy (or the timeout interval) and is also independent of whether I reset the user agent in `didCommitNavigation` or not.

The fact that there is no actual network request when using `loadRequest` could mean that `WKWebView` handles the caching autonomously either on top of the standard URL content cache or in its own cache. I wasn't able to dig up any documentation on this behavior, however, and it's very confusing that it only seems to happen on one particular site.

I made a change to use the "cache-policy initializer" of `NSURLRequest` when the initial URL is different from the current URL. Shall I keep it like this or shall I use `loadRequest` in all cases?

PS: Fully agree on the `m.` subdomains. :)

Johennes commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -1020,6 +1020,24 @@ extension BrowserViewController: BrowserToolbarDelegate {
+
+            if currentItem.URL != currentItem.initialURL {
+                // Reload the initial URL to avoid UA specific redirection
+                tab.loadRequest(NSURLRequest(URL: currentItem.initialURL, cachePolicy: .ReloadIgnoringLocalAndRemoteCacheData, timeoutInterval: 60))
```
I chose the standard timeout of `NSURLRequest` which, according to the [docs](https://developer.apple.com/library/mac/documentation/Cocoa/Reference/Foundation/Classes/NSURLRequest_Class/#//apple_ref/occ/clm/NSURLRequest/requestWithURL:), is 60 seconds.

Johennes commented:
I added the action sheet UI as requested by Darrin on bugzilla.

Johennes commented on Client/Frontend/Browser/Browser.swift:
```diff
@@ -53,6 +53,10 @@ class Browser: NSObject {
 
+    /// The user agent override used for the last request, reload or navigation. Note that this property needs to
+    /// be managed by the web view's navigation delegate.
+    var lastCustomUserAgent: String? = nil
```
I'm not fully satisfied with this state variable being managed outside of the `Browser` class. The latter would need to intercept the `WKNavigationDelegate` methods in order to manage the variable on its own but that appeared to be overly complicated to me.

thebnich replied:
How about just having this be a boolean, `desktopSite`? It doesn't look like the string is actually used anywhere.

Johennes commented on Client/Frontend/Browser/Browser.swift:
```diff
@@ -349,6 +353,27 @@ class Browser: NSObject {
     }
+
+    @available(iOS 9, *)
+    func toggleDesktopSite(on: Bool) {
```
I moved the actual user agent setting and reloading to the `Browser` class since the `BrowserViewController` class is already quite large and this logic could be considered internal to the `Browser`.

Johennes commented on Client/Frontend/Browser/TabManager.swift:
```diff
@@ -327,6 +327,13 @@ class TabManager : NSObject {
         return -1
     }
 
+    func getTab(webView: WKWebView) -> Browser? {
```
Again, I'm not 100% satisfied with this new method but couldn't find a better way. The `WKNavigationDelegate` methods only include the web view but not the tab. We do, however, need the tab in order to store the last user agent override for the action sheet toggling. This cannot be stored on the web view itself without subclassing it.

Johennes replied:
I just realized that this functionality already exists with the `subscript` operator and removed this method.

thebnich commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -1005,6 +1005,31 @@ extension BrowserViewController: BrowserToolbarDelegate {
+        let controller = UIAlertController(title: nil, message: nil, preferredStyle: .ActionSheet)
+        controller.addAction(UIAlertAction(title: toggleActionTitle, style: .Default, handler: { action in tab.toggleDesktopSite(desktopSite) }))
+        controller.addAction(UIAlertAction(title: NSLocalizedString("Cancel", comment:"Action Sheet Cancel Button"), style: .Cancel, handler: nil))
+        presentViewController(controller, animated: true, completion: nil)
```
We'll need to use a `UIPopoverPresentationController` here so that we don't crash on tablets. See `browserToolbarDidPressShare` in this same file for an example.

Johennes replied:
Shoot, I forgot about iPad. Thanks for the reminder. I fixed it.

thebnich commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -1455,6 +1480,16 @@ extension BrowserViewController: TabManagerDelegate {
 }
 
 extension BrowserViewController: WKNavigationDelegate {
+    func webView(webView: WKWebView, didCommitNavigation navigation: WKNavigation!) {
```
Even if `didCommitNavigation` works, `didFinishNavigation` seems more appropriate, I think, since it's called after the main frame load completes. Logically, that's what we would want to wait for before setting the UA for the next page.

Johennes replied:
Yes, that sounds reasonable. I made the change.

Johennes commented:
One more thing I just came to think about is how do we want the normal reload tap to work when the current page was loaded with the desktop UA? Currently it will reload the page with the default mobile UA. Would it maybe be better to instead reload with the UA of the last navigation so that the user is not kicked back to the mobile presentation?

thebnich commented on Client/Frontend/Browser/Browser.swift:
```diff
@@ -356,6 +360,27 @@ class Browser: NSObject {
+        }
+
+        if on {
+            let defaults = NSUserDefaults(suiteName: AppInfo.sharedContainerIdentifier())!
```
It doesn't feel right to have to create an `NSUserDefaults` instance here. Passing the `NSUserDefaults` instance to `UserAgent` actually seems like overkill to me; we could have `UserAgent` create and hold its own `NSUserDefaults` that we don't need to pass into these methods.

Would you mind moving this into `UserAgent`, then removing the `suiteName` parameter from all the methods currently accepting it?

thebnich commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -1015,6 +1015,33 @@ extension BrowserViewController: BrowserToolbarDelegate {
+        let toggleActionTitle: String
+        let desktopSite: Bool
+        if let userAgent = tab.lastCustomUserAgent where !userAgent.isEmpty {
+            toggleActionTitle = NSLocalizedString("Request Mobile Site", comment: "Action Sheet Button for Requesting the Mobile Site")
```
Please remove this string and the string below from `UIConstants`. We have placeholder strings for these in that file so that they'd be exported and localized so that we can ship this feature sooner. Since you're using the strings here, the placeholders aren't needed anymore.

thebnich commented on Client/Frontend/Browser/Browser.swift:
```diff
@@ -356,6 +360,27 @@ class Browser: NSObject {
+            webView.customUserAgent = UserAgent.desktopUserAgent(defaults)
+        } else {
+            webView.customUserAgent = nil
+        }
```
Then this could be reduced to: `webView.customUserAgent = on ? UserAgent.desktopUserAgent() : nil`.

thebnich commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -1015,6 +1015,33 @@ extension BrowserViewController: BrowserToolbarDelegate {
+        }
+
+        let toggleActionTitle: String
+        let desktopSite: Bool
```
Shouldn't need this anymore if you move this boolean to `Browser`.

Johennes commented:
Thanks for the suggestions. I worked everything in. Short summary:

- Deleted placeholder strings from `UIConstants`
- Removed the `defaults` parameter from `desktopUserAgent`. I wasn't sure if you also wanted me to remove the argument from `defaultUserAgent`. I think it would make sense but wanted to check back with you first. In any case, I chose to not use a static variable for the defaults because both methods are called rather infrequently.
- Turned `lastCustomUserAgent` string into a boolean `desktopSite`. I could not perform the toggling in `didSet`, however, because that led to an infinite loop. In `didFinishNavigation` we need to remember whether the desktop site was requested or not (for the toggling). When we set `desktopSite` here and have the toggling in `didSet`, that'll lead to another reload and we start over.

In addition, I added two basic UI tests.

thebnich commented:
> I wasn't sure if you also wanted me to remove the argument from defaultUserAgent. I think it would make sense but wanted to check back with you first.

Yes, please :)

Everything looks great! If you don't mind making that change, I'll go ahead and merge with the updated PR.

Johennes commented:
Ok, it's done. :)

Since the defaults argument in `defaultUserAgent` is needed in `ClientTests.swift`, I kept it around but made it optional with a `nil` default value. When the method is called without an argument, it'll load `NSUserDefaults(suiteName: AppInfo.sharedContainerIdentifier())!` then. I also updated the users of `defaultUserAgent` - the `UserAgent` class itself and the `AppDelegate`.

thebnich commented:
> Since the defaults argument in defaultUserAgent is needed in ClientTests.swift

Is it actually needed, or could we just cut it out like you did here? It looks like `ClientTests.swift` uses `AppInfo.sharedContainerIdentifier()`, which is what we're using everywhere else, right?

Johennes commented:
@thebnich Yes, you're absolutely right. The tests use the very same defaults as the app so there is really no need to keep them injectable. Sorry, I missed that.

I dropped the defaults argument from all methods in `UserAgent` now and instead set up a static variable. I guess it's ok to keep the defaults around like that. `NSUserDefaultsPrefs`, for instance, seems to do the same (but on an instance level).

thebnich commented:
Awesome, thanks a ton for the patch and all of the follow-up fixes! :)

Johennes commented:
Thanks for letting me have it in the first place although it was on your roadmap. This was a great experience! :)

thebnich commented:
Merged into v1.x: 09129e4
